# Тема “Обучение с учителем”

### Задание 1
* Импортируйте библиотеки pandas и numpy.
* Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.
* Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.
* Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.
* Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.
* Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [1]:
import numpy as np
import pandas as pd

In [2]:
from sklearn.datasets import load_boston
boston = load_boston()

In [3]:
data = boston["data"]
feature_names = boston["feature_names"]
target = boston["target"]
X = pd.DataFrame(data, columns=feature_names)
y = pd.DataFrame(target, columns=["price"])

In [4]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42, test_size=0.3)

In [5]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred = lr.predict(X_test)

In [6]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.7112260057484908

**Ответ:** $R^2$ = 0.7112260057484908



### Задание 2

* Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.
* Сделайте агрумент n_estimators равным 1000,
* max_depth должен быть равен 12 и random_state сделайте равным 42.
* Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression, но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0], чтобы получить из датафрейма одномерный массив Numpy, так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.
* Сделайте предсказание на тестовых данных и посчитайте R2.
* Сравните с результатом из предыдущего задания.
* Напишите в комментариях к коду, какая модель в данном случае работает лучше.

In [7]:
from sklearn.ensemble import RandomForestRegressor

In [8]:
model = RandomForestRegressor(random_state=42, n_estimators=1000, max_depth=12)

In [9]:
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=12,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=1000,
                      n_jobs=None, oob_score=False, random_state=42, verbose=0,
                      warm_start=False)

In [10]:
y_pred_rfr = model.predict(X_test)

In [11]:
r2_score(y_test, y_pred_rfr)

0.8749965273218174

**Ответ:** $R^2$ = 0.8749965273218174

### Комментарии:
* Для модели линейной регерессии $R^2$ = 0.7112260057484908
* Для  модели   "случайный лес"  $R^2$ = 0.8749965273218174
* **Для второй модели $R^2$ получился больше, то есть вторая модель предсказывает цены на недвижимость лучше.**

### *Задание 3
* Вызовите документацию для класса RandomForestRegressor,
* найдите информацию об атрибуте feature_importances_.
* С помощью этого атрибута найдите сумму всех показателей важности,
* установите, какие два признака показывают наибольшую важность.

In [12]:
# ?RandomForestRegressor

Справка по атрибуту *feature_importances_*:
    
    feature_importances_ : array of shape = [n_features]        
        The feature importances (the higher, the more important the feature).

In [13]:
feature_importances = model.feature_importances_
feature_importances

array([0.03211748, 0.00154999, 0.0070941 , 0.0011488 , 0.01436832,
       0.40270459, 0.01424477, 0.06403265, 0.00496762, 0.01169177,
       0.01808961, 0.0123114 , 0.41567892])

Сумма всех показателей важности:

In [14]:
feature_importances.sum()

0.9999999999999999

Индексы двух лучших показателей:

In [15]:
bst_ind = feature_importances.argsort()[-2:]
bst_ind

array([ 5, 12], dtype=int64)

Два самых важных признака (с их показателями важности):

In [16]:
two_best_features = dict()
for i in bst_ind:
    two_best_features[feature_names[i]] = feature_importances[i]
two_best_features

{'RM': 0.402704591696731, 'LSTAT': 0.4156789214509943}

Сумма показателей важности двух самых выжных признаков:

In [17]:
sum(two_best_features.values())

0.8183835131477253

**Важность признаков RM и LSTAT составляет 82% от важности всех признаков.**